From e1a81aea269bfa6731a771bf7b8032f40d221a19 Mon Sep 17 00:00:00 2001
From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Mon, 1 Mar 2010 20:52:04 -0300
Subject: [PATCH 02/20] qemu-kvm: enable get/set vcpu events on reset and migration

RH-Author: Marcelo Tosatti <mtosatti@redhat.com>
Message-id: <a5d73e689f3f7fbc6eacb63ced8365b1112e40a8.1267476474.git.mtosatti@redhat.com>
Patchwork-id: 7360
O-Subject: [PATCH 02/21] qemu-kvm: enable get/set vcpu events on reset and
	migration
Bugzilla: 569613
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Juan Quintela <quintela@redhat.com>

qemu-kvm should reset and save/restore vcpu events.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Acked-by: Jan Kiszka <jan.kiszka@siemens.com>
Acked-by: Gleb Natapov <gleb@redhat.com>
(cherry picked from commit ae5ae0d52085d9f16216741c27736cf8fbdfe22d)
---
 kvm.h                 |    5 +++++
 qemu-kvm-x86.c        |    3 ++-
 qemu-kvm.c            |    5 +++++
 target-i386/kvm.c     |    8 ++++++--
 target-i386/machine.c |    6 +++++-
 5 files changed, 23 insertions(+), 4 deletions(-)

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
---
 kvm.h                 |    5 +++++
 qemu-kvm-x86.c        |    3 ++-
 qemu-kvm.c            |    5 +++++
 target-i386/kvm.c     |    8 ++++++--
 target-i386/machine.c |    6 +++++-
 5 files changed, 23 insertions(+), 4 deletions(-)

diff --git a/kvm.h b/kvm.h
index 3a7c2bc..93bce5e 100644
--- a/kvm.h
+++ b/kvm.h
@@ -58,6 +58,9 @@ int kvm_set_migration_log(int enable);
 int kvm_has_sync_mmu(void);
 #endif /* KVM_UPSTREAM */
 int kvm_has_vcpu_events(void);
+int kvm_put_vcpu_events(CPUState *env);
+int kvm_get_vcpu_events(CPUState *env);
+
 #ifdef KVM_UPSTREAM
 
 void kvm_setup_guest_memory(void *start, size_t size);
@@ -102,7 +105,9 @@ int kvm_arch_init(KVMState *s, int smp_cpus);
 
 int kvm_arch_init_vcpu(CPUState *env);
 
+#endif
 void kvm_arch_reset_vcpu(CPUState *env);
+#ifdef KVM_UPSTREAM
 
 struct kvm_guest_debug;
 struct kvm_debug_exit_arch;
diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index 82e362c..7f820a4 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -1457,8 +1457,9 @@ void kvm_arch_push_nmi(void *opaque)
 
 void kvm_arch_cpu_reset(CPUState *env)
 {
-    env->interrupt_injected = -1;
+    kvm_arch_reset_vcpu(env);
     kvm_arch_load_regs(env);
+    kvm_put_vcpu_events(env);
     if (!cpu_is_bsp(env)) {
 	if (kvm_irqchip_in_kernel()) {
 #ifdef KVM_CAP_MP_STATE
diff --git a/qemu-kvm.c b/qemu-kvm.c
index 973695d..f0918b3 100644
--- a/qemu-kvm.c
+++ b/qemu-kvm.c
@@ -2190,6 +2190,11 @@ static int kvm_create_context(void)
         return r;
     }
 
+    kvm_state->vcpu_events = 0;
+#ifdef KVM_CAP_VCPU_EVENTS
+    kvm_state->vcpu_events = kvm_check_extension(kvm_state, KVM_CAP_VCPU_EVENTS);
+#endif
+
     kvm_init_ap();
     if (kvm_irqchip) {
         if (!qemu_kvm_has_gsi_routing()) {
diff --git a/target-i386/kvm.c b/target-i386/kvm.c
index 8584507..e813b73 100644
--- a/target-i386/kvm.c
+++ b/target-i386/kvm.c
@@ -226,12 +226,14 @@ int kvm_arch_init_vcpu(CPUState *env)
     return kvm_vcpu_ioctl(env, KVM_SET_CPUID2, &cpuid_data);
 }
 
+#endif
 void kvm_arch_reset_vcpu(CPUState *env)
 {
     env->interrupt_injected = -1;
     env->nmi_injected = 0;
     env->nmi_pending = 0;
 }
+#ifdef KVM_UPSTREAM
 
 static int kvm_has_msr_star(CPUState *env)
 {
@@ -716,8 +718,9 @@ static int kvm_get_mp_state(CPUState *env)
     env->mp_state = mp_state.mp_state;
     return 0;
 }
+#endif
 
-static int kvm_put_vcpu_events(CPUState *env)
+int kvm_put_vcpu_events(CPUState *env)
 {
 #ifdef KVM_CAP_VCPU_EVENTS
     struct kvm_vcpu_events events;
@@ -747,7 +750,7 @@ static int kvm_put_vcpu_events(CPUState *env)
 #endif
 }
 
-static int kvm_get_vcpu_events(CPUState *env)
+int kvm_get_vcpu_events(CPUState *env)
 {
 #ifdef KVM_CAP_VCPU_EVENTS
     struct kvm_vcpu_events events;
@@ -784,6 +787,7 @@ static int kvm_get_vcpu_events(CPUState *env)
     return 0;
 }
 
+#ifdef KVM_UPSTREAM
 int kvm_arch_put_registers(CPUState *env)
 {
     int ret;
diff --git a/target-i386/machine.c b/target-i386/machine.c
index 47ca6e8..0b8a33a 100644
--- a/target-i386/machine.c
+++ b/target-i386/machine.c
@@ -323,7 +323,10 @@ static void cpu_pre_save(void *opaque)
     int i;
 
     cpu_synchronize_state(env);
-    kvm_save_mpstate(env);
+    if (kvm_enabled()) {
+        kvm_save_mpstate(env);
+        kvm_get_vcpu_events(env);
+    }
 
     /* FPU */
     env->fpus_vmstate = (env->fpus & ~0x3800) | (env->fpstt & 0x7) << 11;
@@ -374,6 +377,7 @@ static int cpu_post_load(void *opaque, int version_id)
 
         kvm_load_tsc(env);
         kvm_load_mpstate(env);
+        kvm_put_vcpu_events(env);
     }
 
     return 0;
-- 
1.7.0.3

